SourceCode/raytracer/raytracer/Renderer.h

```diff
@@ -0,0 +1,153 @@
+#pragma once
+#include <algorithm>
+#include <tuple>
+#include <queue>
+
+#include "Camera.h"
+#include "CRTTypes.h"
+#include "Image.h"
+#include "Scene.h"
+#include "Triangle.h"
+
+#include "perlin-noise.h"
+
+class Renderer {
+public:
+    Renderer() {}
+
+    void renderScene(const Scene& scene, Image& image)
+    {
+        // Prepare Primary Queue
+        scene.metrics.startTimer();
+        Ray ray;
+        for (int y = 0; y < image.getHeight(); ++y) {
+            for (int x = 0; x < image.getWidth(); ++x) {
+                scene.camera.emplacePrimaryRay(image, x, y, primaryQueue);
+            }
+        }
+
+        processPrimaryQueue(scene, image);
+        processShadowQueue(scene, image);
```

Comment:
Probably you've fixed this in the next PRs, but having a separate shadow queue has issues in some of your rendered images shadows are not generate below objects